File: datasets/Python/Mistral/DataStructures/easy4.py
def getMax(operations):
    stack = []
    max_stack = []
    output = []

    for operation in operations:
        parts = operation.split()
        if parts[0] == '1':
            x = int(parts[1])
            stack.append(x)
            if not max_stack or x >= max_stack[-1]:
                max_stack.append(x)
        elif parts[0] == '2':
            if stack:
                popped = stack.pop()
                if max_stack and popped == max_stack[-1]:
                    max_stack.pop()
        elif parts[0] == '3':
            if max_stack:
                output.append(max_stack[-1])

    return output

def main():
    n = int(input("Enter the number of operations: "))
    operations = []
    for _ in range(n):
        operations.append(input("Enter operation: "))
    results = getMax(operations)
    for result in results:
        print(result)

if __name__ == "__main__":
    main()

File: datasets/C++/Mistral/DataStructures/easy4.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>

using namespace std;

vector<int> getMax(const vector<string>& operations) {
    vector<int> stack;
    vector<int> maxStack;
    vector<int> output;

    for (const string& operation : operations) {
        istringstream iss(operation);
        int type, x;
        iss >> type;
        if (type == 1) {
            iss >> x;
            stack.push_back(x);
            if (maxStack.empty() || x >= maxStack.back()) {
                maxStack.push_back(x);
            }
        } else if (type == 2) {
            if (!stack.empty()) {
                int popped = stack.back();
                stack.pop_back();
                if (!maxStack.empty() && popped == maxStack.back()) {
                    maxStack.pop_back();
                }
            }
        } else if (type == 3) {
            if (!maxStack.empty()) {
                output.push_back(maxStack.back());
            }
        }
    }

    return output;
}

int main() {
    int n;
    cout << "Enter the number of operations: ";
    cin >> n;
    cin.ignore(); // Ignore the newline character after reading n

    vector<string> operations(n);
    for (int i = 0; i < n; ++i) {
        cout << "Enter operation: ";
        getline(cin, operations[i]);
    }

    vector<int> results = getMax(operations);
    for (int result : results) {
        cout << result << endl;
    }

    return 0;
}
